lib/scholar/metrics/distance.ex

```diff
@@ -198,17 +284,54 @@ defmodule Scholar.Metrics.Distance do
       iex> y = Nx.tensor([1, 2, 3])
       iex> Scholar.Metrics.Distance.minkowski(x, y)
       ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+
+      iex> x = Nx.tensor([[1,2], [3,4]])
+      iex> y = Nx.tensor([[8,3], [2,5]])
+      iex> Scholar.Metrics.Distance.minkowski(x, y, p: 2.5, axes: [0])
+      #Nx.Tensor<
+        f32[2]
+        [7.021548271179199, 1.3195079565048218]
+      >
   """
-  @spec minkowski(Nx.t(), Nx.t(), integer()) :: Nx.t()
-  defn minkowski(x, y, p \\ 2) do
+  @spec minkowski(Nx.t(), Nx.t()) :: Nx.t()
+  defn minkowski(x, y, opts \\ []) do
     assert_same_shape!(x, y)
 
-    x
-    |> Nx.subtract(y)
-    |> Nx.abs()
-    |> Nx.power(p)
-    |> Nx.sum()
-    |> Nx.power(1.0 / p)
+    opts =
+      keyword!(
+        opts,
+        p: 2,
+        axes: nil
+      )
+
+    p = Nx.tensor(opts[:p])
+    check_p(p)
+
+    case p do
+      0 ->
+        chebyshev(x, y, axes: opts[:axes])
+
+      1 ->
+        manhattan(x, y, axes: opts[:axes])
+
+      2 ->
+        euclidean(x, y, axes: opts[:axes])
+
+      _ ->
+        x
+        |> Nx.subtract(y)
+        |> Nx.abs()
+        |> Nx.power(p)
+        |> Nx.sum(axes: opts[:axes])
+        |> Nx.power(1.0 / p)
+    end
+  end
+
+  deftransformp check_p(p) do
+    if p < 0 do
```

Comment:
I don't think this check is working as intended given that `p` is passed as a tensor to the check

Reply:
@polvalente I think on latest Nx this shouldn't matter because if works at compile time if the condition value is a constant. That said, I don't think there's a reason to explicitly convert to a tensor.